proposals/XXXX-associated-types-constraints.md

```diff
@@ -0,0 +1,75 @@
+# Permit where clauses to constrain associated types
+
+* Proposal: [SE-XXXX](https://github.com/apple/swift-evolution/blob/master/proposals/XXXX-associated-types-constraints.md)
+* Author(s): [David Hart](http://github.com/hartbit), [Jacob Bandes-Storch](jtbandes@gmail.com), [Douglas Gregor](dgregor@apple.com)
+* Status: TBD
+* Review manager: TBD
+
+## Introduction
+
+This proposal seeks to introduce a `where` clause to associated types declarations and improvements to protocol constraints to bring associated types the same expressive power as generic types.
+
+This proposal was discussed twice on the Swift Evolution list in the following threads:
+
+- [[Completing Generics] Arbitrary requirements in protocols](http://thread.gmane.org/gmane.comp.lang.swift.evolution/14243)
+- [More Powerful Constraints for Associated Types](http://thread.gmane.org/gmane.comp.lang.swift.evolution/15201)
+ 
+## Motivation
+
+Currently, associated type declarations can only express simple inheritance constraints and not the more sophisticated constraints available to generic types with the `where` clause. Some designs, including many in the Standard Library, require more powerful constraints for associated types to be truly elegant. For example, the `SequenceType` protocol can be declared as follows:
+
+```swift
+protocol Sequence {
+    associatedtype Iterator : IteratorProtocol
+    associatedtype SubSequence : Sequence where SubSequence.Iterator.Element == Iterator.Element
+    ...
+}
+```
+
+## Detailed Design
+
+First of all, this proposal modifies the grammar for protocols associated types to the following:
+
+*protocol-associated-type-declaration* → *attributes<sub>opt</sub>* *access-level-modifier<sub>opt</sub>* **associatedtype** *typealias-name* ­*type-inheritance-clause­<sub>opt</sub>* *typealias-assignment­<sub>opt</sub>* *requirement-clause<sub>opt</sub>*
+
+The new requirement-clause is then used by the compiler to validate the associated types of conforming types.
+
+Secondly, the proposal also allows protocols to use the associated types of their conforming protocols in their declaration where clause as below:
+
+```swift
+protocol IntSequence : Sequence where Iterator.Element == Int {
+    ...
+}
+```
+
+Name lookup semantics in the protocol declaration `where` clause only looks at associated types in the parent protocols. For example, the following code would cause an error:
+
+```swift
+protocol IntSequence : Sequence where Counter: Int { // error: Use of undefined associated type 'Counter'
```

Comment:
s/`Counter: Int`/`Counter : SomeProtocol`/?

Constraining counter to be `Int` exactly I think makes it a non-associated type, better modeled by a typealias in the protocol.